include/permissions.hxx

```diff
@@ -0,0 +1,216 @@
+  /// Placeholders can be used in variable names by surrounding a
+  /// label with curly braces. Multiple values can then be substituted
+  /// for this placeholder using the Permissions::substitute
+  /// method. For example, if you wanted to read the collision
+  /// frequency for every species you would write:
+  ///
+  ///     Permissions example2({
+  ///         {"species:{name}:collision_frequency", {Permissions::Nowhere,
+  ///         Permissions::AllRegions, Permissions::Nowhere, Permissions::Nowhere}}
+  ///     });
+  ///     example2.substitute("name", {"he+", "d+", "e", "d", "he"});
+  ///
+  Permissions(std::initializer_list<std::pair<std::string, AccessRights>> data);
+
+  /// Set the level of access for the various regions of the
+  /// variable. This uses the same logic as the constructor. For
+  /// example, to indicate that the density of helium is readable
+  /// everywhere but only writeable in the interior, you would use
+  ///
+  ///     permissions.setAccess("species:he:density",
+  ///                           {Permissions::Nowhere, Permissions::AllRegions,
+  ///                           Permissions::Interior, Permissions::Nowhere})
+  /// 0
+  /// or, equivalently,
+  ///
+  ///     permissions.setAccess("species:he:density",
+  ///                           {Permissions::Nowhere, Permissions::Boundary,
+  ///                           Permissions::Interior, Permissions::Nowhere});
+  ///
+  /// As in the constructor, if the variable name is just a section in
+  /// an Options object then the permissions apply to all children of
+  /// that section. Placeholder names can also be used.
+  void setAccess(const std::string& variable, const AccessRights& rights);
+
+  /// Replace a placeholder in the names of variables stored in this
+  /// object. This is useful if you need to access the same variable
+  /// for multiple species. For example, the following code gives
+  /// permission to read the density and write the collision frequency
+  /// for every species.
+  ///
+  ///     Permissions example({
+  ///         {"species:{name}:density", {Permissions::Nowhere, Permissions::AllRegions,
+  ///         Permissions::Nowhere, Permissions::Nowhere}},
+  ///         {"species:{name}:collision_frequency", {Permissions::Nowhere,
+  ///         Permissions::Nowhere, Permissions::AllRegions, Permissions::Nowhere}},
+  ///     });
+  ///     example.substitute("name", {"d", "d+", "t", "t+", "he", "he+", "c", "c+", "e"});
+  ///
+  void substitute(const std::string& label,
+                  const std::vector<std::string>& substitutions);
+
+  /// Check whether users are allowed to access this variable to the
+  /// given permission level, in the given region. The second item
+  /// returned indicates the name of the variable or section from
+  /// which the access rights are derived. If there is no matching
+  /// section then it will be an empty string.
+  std::pair<bool, std::string> canAccess(const std::string& variable,
```

Comment:
Is this return type basically `std::optional<std::string>`? Might be cleaner to switch to that?

Reply:
You are right that this would express the same information. I'm not sure whether that's the best approach though. The main information being returned here is whether the access rights exist. The variable name from which those rights are derived is secondary information. I'm not sure it would make sense to effectively make that the primary information by returning a `std::optional<std::string>`.